### PR TITLE
unit tests and fixes

### PR DESCRIPTION
**Pull request by omershlo**
- #7 
- added commitment without user defined randomness 
- made some changes to party_one.rs (sample bits instead of bytes)

### CODE REVIEW

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 1]
```diff
@@ -33,20 +34,55 @@ impl Commitment for HashCommitment {
...
         BigInt::from(digest.finish().as_ref())
     }
+
+    fn create_commitment(
+        message: &BigInt, security_bits: &usize) -> (BigInt, BigInt)
+    {
+        let mut digest = Context::new(&SHA256);
+        let bytes_message: Vec<u8> = message.into();
+        digest.update(&bytes_message);
+        let blinding_factor = &(BigInt::sample(security_bits.clone()));
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: add empty line above

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 2]
```diff
@@ -33,20 +34,55 @@ impl Commitment for HashCommitment {
...
-        let commitment = HashCommitment::create_commitment_with_user_defined_randomness(
-            &BigInt::one(), &BigInt::zero());
+    fn hash_commitment_test() {
+        let sec_bits = 256;
+        let message = BigInt::sample(sec_bits.clone());
+        let (commitment, blind_factor) = HashCommitment::create_commitment(&message, &sec_bits);
+        let commitment2 = HashCommitment::create_commitment_with_user_defined_randomness(
+            &message, &blind_factor);
+        //test commitment length  - works because SHA256 output length the same as sec_bits
+        assert_eq!(commitment.bit_length(),sec_bits);
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: Can we create 2 tests, one for create_commitment and one for create_commitment_with_user_defined_randomness that will have a name which self explained the purpose: "create_commitment_bit_length_test" and "create_commitment_with_user_defined_randomness_bit_length_test" and move the code accordingly. It is always better to localized your unit tests as much as you can.

Also I suggest to remove print statements (even if commented).

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 3]
```diff
@@ -33,20 +34,55 @@ impl Commitment for HashCommitment {
...
+        assert_eq!(commitment, commitment2);
+        // debug:
+        //println!("commitment: {:?}", commitment.to_str_radix(16));
+        //println!("length: {:?}", commitment.bit_length());
+    }
 
-        println!("{}", commitment);
+    #[test]
+    fn hash_test() {
+
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: remove empty line

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 4]
```diff
@@ -33,20 +34,55 @@ impl Commitment for HashCommitment {
...
+        //test commitment correctness
+        assert_eq!(commitment, commitment2);
+        // debug:
+        //println!("commitment: {:?}", commitment.to_str_radix(16));
+        //println!("length: {:?}", commitment.bit_length());
+    }
 
-        println!("{}", commitment);
+    #[test]
+    fn hash_test() {
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: Can we find a better name? create_commitment_with_user_defined_randomness_hash_test

File: src/cryptographic_primitives/commitments/traits.rs [hunk 5]
```diff
@@ -19,4 +19,7 @@ use ::BigInteger as BigInt;
 pub trait Commitment {
     fn create_commitment_with_user_defined_randomness(
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/traits.rs]: Maybe we need to add a // TODO: add documentation (and then create an issue)

**Comment by omershlo** [comment]
@gbenattar the new commit answer your review. Thanks.

I found a problem using the unit test (I actually found a few but only one I couldn't fix): 
Because the input is a bigint I can't commit to an empty message or use empty randomness.
If the inputs were strings I could simply use "" for empty string. Do you have a way to do it using BigInt?

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 6]
```diff
@@ -33,20 +36,73 @@ impl Commitment for HashCommitment {
...
+        println!("blind_factor: {:?}", blind_factor.to_str_radix(16));
+        assert_eq!(commitment.to_str_radix(16).len(),SECURITY_BITS/4);
+        assert_eq!(blind_factor.to_str_radix(16).len(),SECURITY_BITS/4);
 
+
+    }
+
+    #[test]
+    fn test_bit_length_create_commitment_with_user_defined_randomness() {
+
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: remove empty line.

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 7]
```diff
@@ -33,20 +36,73 @@ impl Commitment for HashCommitment {
...
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use ::BigInteger as BigInt;
     use super::Commitment;
     use super::HashCommitment;
+    use arithmetic::traits::Samplable;
+    const SECURITY_BITS : usize = 256;
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: Can you extract this at the mod.rs level. Then you can use it in both unit test and implementation

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 8]
```diff
@@ -33,20 +36,73 @@ impl Commitment for HashCommitment {
...
+    #[test]
+    fn test_bit_length_create_commitment() {
+        let message = BigInt::sample(SECURITY_BITS);
+        let (commitment, blind_factor) = HashCommitment::create_commitment(&message);
+        //test commitment length  - works because SHA256 output length the same as sec_bits
+        println!("blind_factor: {:?}", blind_factor.to_str_radix(16));
+        assert_eq!(commitment.to_str_radix(16).len(),SECURITY_BITS/4);
+        assert_eq!(blind_factor.to_str_radix(16).len(),SECURITY_BITS/4);
 
+
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: remove empty line

File: src/cryptographic_primitives/commitments/hash_commitment.rs [hunk 9]
```diff
@@ -33,20 +36,73 @@ impl Commitment for HashCommitment {
...
     use super::HashCommitment;
+    use arithmetic::traits::Samplable;
+    const SECURITY_BITS : usize = 256;
+
+    #[test]
+    fn test_bit_length_create_commitment() {
+        let message = BigInt::sample(SECURITY_BITS);
+        let (commitment, blind_factor) = HashCommitment::create_commitment(&message);
+        //test commitment length  - works because SHA256 output length the same as sec_bits
+        println!("blind_factor: {:?}", blind_factor.to_str_radix(16));
```
Comment (gbenattar) [inline src/cryptographic_primitives/commitments/hash_commitment.rs]: remove println

**Comment by gbenattar** [comment]
Maybe @mortendahl can help about the question related to empty randomness (see above).

**Comment by omershlo** [comment]
We don't need to at the moment. Probably committing without randomness is more likely.

There are commitment schemes where randomness r=0 is the same as not using randomness (Pedersen) where in this case using r=0 is not equal to not use r at all. This was my main point.

I don't believe it's a big issue. If in the future we will need to use a hash commitment without randomness we will just add a function with only one element as input.

**Comment by omershlo** [comment]
@gbenattar answered your review in latest commit

**Pull request merged**